samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java

```diff
@@ -156,4 +161,46 @@ public ContextManager getContextManager() {
   /* package private */ int getNextOpId() {
     return this.opId++;
   }
+
+  /**
+   * Get all {@link OperatorSpec}s available in this {@link StreamGraphImpl}
+   *
+   * @return  a set of all available {@link OperatorSpec}s
+   */
+  public Set<OperatorSpec> getAllOperatorSpecs() {
+    Collection<InputStreamInternal> inputStreams = inStreams.values();
+    Set<OperatorSpec> operatorSpecs = new HashSet<>();
+
+    for (InputStreamInternal stream : inputStreams) {
+      doGetOperatorSpecs((MessageStreamImpl) stream, operatorSpecs);
+    }
+    return operatorSpecs;
+  }
+
+  private void doGetOperatorSpecs(MessageStreamImpl stream, Set<OperatorSpec> specs) {
+    Collection<OperatorSpec> registeredOperatorSpecs = stream.getRegisteredOperatorSpecs();
+    for (OperatorSpec spec : registeredOperatorSpecs) {
+      specs.add(spec);
+      MessageStreamImpl nextStream = spec.getNextStream();
+      if (nextStream != null) {
+        //Recursively traverse and obtain all reachable operators
+        doGetOperatorSpecs(nextStream, specs);
+      }
+    }
+  }
+
+  /**
+   * Returns <tt>true</tt> iff this {@link StreamGraphImpl} contains a join or a window operator
+   *
+   * @return  <tt>true</tt> iff this {@link StreamGraphImpl} contains a join or a window operator
+   */
+  public boolean hasWindowOrJoins() {
```

Comment:
nit: window and join are interesting not because of their identities, but because of their behavior; that they both depend on the "window" phase of the event loop. If there's an obvious name that captures this, it would be better. 

So, something like "hasTimedOperators" or something that captures the general behavior rather than the current known examples. If there is no such name in this case, feel free to ignore.